spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java

```diff
@@ -0,0 +1,136 @@
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.mongodb.core.query.Update;
+import org.springframework.integration.metadata.ConcurrentMetadataStore;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import com.mongodb.WriteResult;
+
+/**
+ * MongoDbMetadataStore implementation of {@link ConcurrentMetadataStore}.
+ * Use this {@link org.springframework.integration.metadata.MetadataStore} to
+ * achieve meta-data persistence shared across application instances and
+ * restarts.
+ *
+ * @author SenthilArumugam SamirajPanneerSelvam
+ * @since 4.0
+ *
+ */
+public class MongoDbMetadataStore implements ConcurrentMetadataStore {
+
+	private MongoTemplate template;
+	private final static String DEFAULT_COLLECTION_NAME = "metadatastore";
+	private String collectionName;
+
+	/**
+	 * 
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template
+	 * @param collectionName
+	 */
+	public MongoDbMetadataStore(MongoTemplate template, String collectionName) {
+		this.template = template;
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
+				: DEFAULT_COLLECTION_NAME;
+	}
+
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template
+	 */
+	public MongoDbMetadataStore(MongoTemplate template) {
+		this(template, null);
+	}
+
+	@Override
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore fileInfo = new Datastore(key, value);
+		if (!template.exists(query, collectionName)) {
+			template.save(fileInfo, collectionName);
+		}
+	}
+
+	@Override
+	public String get(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore result = template.findOne(query, Datastore.class, collectionName);
+		if(result != null){
+			return result.value;	
+		}
+		return null;
+		
+	}
+
+	@Override
+	public String remove(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore result = template.findAndRemove(query, Datastore.class,collectionName);
+		return result==null?null:result.value;
+	}
+
+	@Override
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+
+		String result = value;
+		Query query = new Query(Criteria.where("key").is(key));
+		if (template.exists(query, collectionName)) {
+			result = null;
+		} else {
+			template.save(new Datastore(key, value), collectionName);
+		}
+		return result;
```

Comment:
From other side this logic doesn't follow with `putIfAbsent` contract:
```
 * If the specified key is not already associated
 * with a value, associate it with the given value.
 * This is equivalent to
 *  <pre> {@code
 * if (!map.containsKey(key))
 *   return map.put(key, value);
 * else
 *   return map.get(key);
 * }</pre>
 *
 * except that the action is performed atomically.
```

Unfortunately I don't see any similar MongoDB atomic operation. So we might end up with something like:
````java
try {
     template.insert(entry, collectionName);
}
catch (MongoDataIntegrityViolationException e)  {
    return get(key);
}
return null;
````
CC @christophstrobl